### PR TITLE
Fix slow spectra

### PR DESCRIPTION
**Pull request by PennyQ**
This PR will fix the slow extraction of spectrum tool problem, as in #1021

### CODE REVIEW

**Comment by PennyQ** [comment]
@astrofrog I found I had a very old commit as 'fix the d3po isinstance error' within this PR, can you show me how to 'safely' remove it? (I made a mess when I tried to rebase)

**Comment by astrofrog** [comment]
@PennyQ - I can help :) Can you show me the output of ``git remove -v``?

**Comment by PennyQ** [comment]
> ➜  glue git:(fix_slow_spectra) git remote -v
origin	https://github.com/PennyQ/glue.git (fetch)
origin	https://github.com/PennyQ/glue.git (push)
upstream	https://github.com/glue-viz/glue.git (fetch)
upstream	https://github.com/glue-viz/glue.git (push)

**Comment by astrofrog** [comment]
@PennyQ - try the following:

```
git fetch upstream
git rebase -i upstream/master
# At this point, you will see a list of commits, you can remove the line for the one that you want to remove. Then save, and the rebase will continue
git push -f origin fix_slow_spectra
```

File: glue/plugins/tools/spectrum_tool/qt/spectrum_tool.py [hunk 1]
```diff
@@ -26,6 +26,8 @@
 from glue.utils.qt import Worker, messagebox_on_error
 from glue.core import roi as core_roi
 
+import time
+
```
Comment (astrofrog) [inline glue/plugins/tools/spectrum_tool/qt/spectrum_tool.py]: I think these changes aren't needed?

File: glue/plugins/tools/spectrum_tool/qt/spectrum_tool.py [hunk 2]
```diff
@@ -68,18 +70,6 @@ def spectrum(data, attribute, roi, slc, zaxis):
...
-            for i in reversed(list(range(ndim))):
-                if i != zaxis:  # index of the zaxis in dataset
-                    data = np.nansum(data, axis=i)
-                    finite = finite.sum(axis=i)
-
-            assert data.ndim == 1
-            assert data.size == nz
-
-            data = (1. * data / finite).ravel()
-
```
Comment (astrofrog) [inline glue/plugins/tools/spectrum_tool/qt/spectrum_tool.py]: Is it normal that you removed all this? Does it not break rectangular spectrum extraction?

File: glue/plugins/tools/spectrum_tool/qt/spectrum_tool.py [hunk 3]
```diff
@@ -93,25 +83,38 @@ def spectrum(data, attribute, roi, slc, zaxis):
...
-                finite = np.isfinite(select_data).sum()
-                return_data.append(1. * np.nansum(select_data) / finite)
+        try:
+            assert data.ndim == ndim
 
-            data = np.array(return_data)
+            for i in reversed(list(range(ndim))):
+                if i != zaxis:  # index of the zaxis in dataset
+                    data = np.nansum(data, axis=i)
+                    finite = finite.sum(axis=i)
```
Comment (astrofrog) [inline glue/plugins/tools/spectrum_tool/qt/spectrum_tool.py]: Just to make sure I understand, does this change the behavior of the spectrum extractor for e.g. 4D cubes? (did it work before?)

File: glue/plugins/tools/spectrum_tool/qt/spectrum_tool.py (hunk 2 above)
Comment (PennyQ) [inline glue/plugins/tools/spectrum_tool/qt/spectrum_tool.py]: I actually move them to the bottom  line 110-115, coz both `Rectangle` and `Circle/Polygon' mode use them to return data.

**Comment by PennyQ** [comment]
This should fix both the axis order problem and 4th dim problem :) @astrofrog

File: glue/plugins/tools/spectrum_tool/qt/spectrum_tool.py [hunk 4]
```diff
@@ -68,50 +68,52 @@ def spectrum(data, attribute, roi, slc, zaxis):
...
-            assert data.ndim == 1
-            assert data.size == nz
+        for i in reversed(list(range(ndim))):
+            if i != zaxis:  # index of the zaxis in dataset
+                data = np.nansum(data, axis=i)
+                finite = finite.sum(axis=i)
 
-        else:
-
-            raise TypeError("Unexpected ROI type: {0}".format(type(roi)))
```
Comment (PennyQ) [inline glue/plugins/tools/spectrum_tool/qt/spectrum_tool.py]: I personally think this kind of Error won't appear so I moved it, coz this line can only be called when the `if ROI.instance` passed, but if later we got similar errors we could add it back ?

**Comment by astrofrog** [comment]
@PennyQ - I'm closing this in favor of #1032 (which includes the changes here and more)

**Pull request closed without merging**